File: src/test/java/util/TrickRulesTest.java
package util;

import com.aau.wizard.model.Card;
import com.aau.wizard.model.Player;
import com.aau.wizard.model.enums.CardSuit;
import com.aau.wizard.model.enums.CardType;
import com.aau.wizard.util.Pair;
import com.aau.wizard.util.TrickRules;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class TrickRulesTest {

    private Player createTestPlayer(String id, String name) {
        return new Player(id, name);
    }

    @Test
    void determineTrickWinner_emptyList_throwsException() {
        assertThrows(IllegalArgumentException.class,
                () -> TrickRules.determineTrickWinner(List.of(), CardSuit.RED));
    }

    @Test
    void determineTrickWinner_singleWizard_wins() {
        Player player = createTestPlayer("p1", "Player1");
        Card wizard = new Card(CardSuit.RED, 14); // Wizard
        List<Pair<Player, Card>> playedCards = List.of(new Pair<>(player, wizard));

        Player winner = TrickRules.determineTrickWinner(playedCards, CardSuit.BLUE);
        assertEquals(player, winner);
    }

    @Test
    void determineTrickWinner_singleJester_winsByDefault() {
        Player player = createTestPlayer("p1", "Player1");
        Card jester = new Card(CardSuit.RED, 0); // Jester
        List<Pair<Player, Card>> playedCards = List.of(new Pair<>(player, jester));

        Player winner = TrickRules.determineTrickWinner(playedCards, CardSuit.BLUE);
        assertEquals(player, winner);
    }

    @Test
    void determineTrickWinner_wizardBeatsAll() {
        Player player1 = createTestPlayer("p1", "Player1");
        Player player2 = createTestPlayer("p2", "Player2");
        Player player3 = createTestPlayer("p3", "Player3");

        List<Pair<Player, Card>> playedCards = List.of(
                new Pair<>(player1, new Card(CardSuit.RED, 7)),
                new Pair<>(player2, new Card(CardSuit.BLUE, 10)),
                new Pair<>(player3, new Card(CardSuit.GREEN, 14)) // Wizard
        );

        Player winner = TrickRules.determineTrickWinner(playedCards, CardSuit.BLUE);
        assertEquals(player3, winner);
    }

    @Test
    void determineTrickWinner_trumpBeatsLeadSuit() {
        Player player1 = createTestPlayer("p1", "Player1");
        Player player2 = createTestPlayer("p2", "Player2");

        List<Pair<Player, Card>> playedCards = List.of(
                new Pair<>(player1, new Card(CardSuit.RED, 10)),
                new Pair<>(player2, new Card(CardSuit.BLUE, 7)) // BLUE is trump
        );

        Player winner = TrickRules.determineTrickWinner(playedCards, CardSuit.BLUE);
        assertEquals(player2, winner);
    }

    @Test
    void determineTrickWinner_higherValueWinsInSameSuit() {
        Player player1 = createTestPlayer("p1", "Player1");
        Player player2 = createTestPlayer("p2", "Player2");

        List<Pair<Player, Card>> playedCards = List.of(
                new Pair<>(player1, new Card(CardSuit.RED, 7)),
                new Pair<>(player2, new Card(CardSuit.RED, 10))
        );

        Player winner = TrickRules.determineTrickWinner(playedCards, CardSuit.BLUE);
        assertEquals(player2, winner);
    }

    @Test
    void determineTrickWinner_jesterDoesntAffectLeadSuit() {
        Player player1 = createTestPlayer("p1", "Player1");
        Player player2 = createTestPlayer("p2", "Player2");
        Player player3 = createTestPlayer("p3", "Player3");

        List<Pair<Player, Card>> playedCards = List.of(
                new Pair<>(player1, new Card(CardSuit.RED, 0)), // Jester
                new Pair<>(player2, new Card(CardSuit.BLUE, 7)),
                new Pair<>(player3, new Card(CardSuit.BLUE, 10))
        );

        Player winner = TrickRules.determineTrickWinner(playedCards, CardSuit.GREEN);
        assertEquals(player3, winner);
    }

    @Test
    void isValidPlay_wizardAlwaysValid() {
        Player player = createTestPlayer("p1", "Player1");
        Card wizard = new Card(CardSuit.RED, 14);

        assertTrue(TrickRules.isValidPlay(player, wizard, List.of()));
        assertTrue(TrickRules.isValidPlay(player, wizard,
                List.of(new Pair<>(player, new Card(CardSuit.BLUE, 7)))));
    }

    @Test
    void isValidPlay_jesterAlwaysValid() {
        Player player = createTestPlayer("p1", "Player1");
        Card jester = new Card(CardSuit.RED, 0);

        assertTrue(TrickRules.isValidPlay(player, jester, List.of()));
        assertTrue(TrickRules.isValidPlay(player, jester,
                List.of(new Pair<>(player, new Card(CardSuit.BLUE, 7)))));
    }

    @Test
    void isValidPlay_firstPlayAlwaysValid() {
        Player player = createTestPlayer("p1", "Player1");
        Card card = new Card(CardSuit.RED, 7);

        assertTrue(TrickRules.isValidPlay(player, card, List.of()));
    }

    @Test
    void isValidPlay_mustFollowSuitIfPossible() {
        Player player = createTestPlayer("p1", "Player1");
        player.setHandCards(List.of(
                new Card(CardSuit.RED, 7),
                new Card(CardSuit.BLUE, 8),
                new Card(CardSuit.GREEN, 9)
        ));

        List<Pair<Player, Card>> currentTrick = List.of(
                new Pair<>(createTestPlayer("p2", "Player2"), new Card(CardSuit.RED, 10))
        );

        assertTrue(TrickRules.isValidPlay(player, new Card(CardSuit.RED, 5), currentTrick));

        assertFalse(TrickRules.isValidPlay(player, new Card(CardSuit.BLUE, 5), currentTrick));

        player.setHandCards(List.of(new Card(CardSuit.BLUE, 8)));
        assertTrue(TrickRules.isValidPlay(player, new Card(CardSuit.BLUE, 5), currentTrick));
    }
}

File: src/test/kotlin/com/aau/wizard/core/rules/TrickRulesTest.kt
package com.aau.wizard.core.rules

import com.aau.wizard.core.cards.*
import com.aau.wizard.core.model.PlayerState
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.assertThrows

class TrickRulesTest {
    private val alice = PlayerState("p1", "Alice")
    private val bob = PlayerState("p2", "Bob")
    private val charlie = PlayerState("p3", "Charlie")


    @Nested
    inner class BasicTrickTests {
        @Test
        fun `throws when no cards played`() {
            assertThrows<IllegalArgumentException> {
                TrickRules.determineTrickWinner(emptyList(), trumpSuit = null)
            }
        }

        @Test
        fun `highest card of lead suit wins with no trump`() {
            val playedCards = listOf(
                alice to Card(Suit.RED, 5),
                bob to Card(Suit.RED, 10),
                charlie to Card(Suit.RED, 7)
            )

            assertEquals(bob, TrickRules.determineTrickWinner(playedCards, trumpSuit = null))
        }

        @Test
        fun `trump card beats lead suit`() {
            val playedCards = listOf(
                alice to Card(Suit.RED, 10),
                bob to Card(Suit.GREEN, 5),
                charlie to Card(Suit.BLUE, 7)
            )

            assertEquals(charlie, TrickRules.determineTrickWinner(playedCards, trumpSuit = Suit.BLUE))
        }
    }

    @Nested
    inner class SpecialCardTests {
        @Test
        fun `wizard always wins`() {
            val playedCards = listOf(
                alice to Card(Suit.RED, 13),
                bob to Card(Suit.SPECIAL, 14),
                charlie to Card(Suit.GREEN, 1)
            )

            assertEquals(bob, TrickRules.determineTrickWinner(playedCards, trumpSuit = Suit.RED))
        }

        @Test
        fun `jester always loses unless all jesters`() {
            val playedCards = listOf(
                alice to Card(Suit.SPECIAL, 0),
                bob to Card(Suit.RED, 2),
                charlie to Card(Suit.GREEN, 1)
            )

            assertEquals(bob, TrickRules.determineTrickWinner(playedCards, trumpSuit = null))
        }

        @Test
        fun `all jesters returns first player`() {
            val playedCards = listOf(
                alice to Card(Suit.SPECIAL, 0),
                bob to Card(Suit.SPECIAL, 0),
                charlie to Card(Suit.SPECIAL, 0)
            )

            assertEquals(alice, TrickRules.determineTrickWinner(playedCards, trumpSuit = Suit.RED))
        }
    }

    @Nested
    inner class LeadSuitTests {
        @Test
        fun `first non-jester determines lead suit`() {
            val playedCards = listOf(
                alice to Card(Suit.SPECIAL, 0),
                bob to Card(Suit.YELLOW, 5),
                charlie to Card(Suit.YELLOW, 7)
            )

            assertEquals(charlie, TrickRules.determineTrickWinner(playedCards, trumpSuit = null))
        }

        @Test
        fun `wizard as first card means no lead suit`() {
            val playedCards = listOf(
                alice to Card(Suit.SPECIAL, 14),
                bob to Card(Suit.RED, 10),
                charlie to Card(Suit.GREEN, 7)
            )

            assertEquals(alice, TrickRules.determineTrickWinner(playedCards, trumpSuit = null))
        }

        @Test
        fun `player must follow lead suit if possible`() {
            val player = PlayerState(
                playerId = "player1",
                name = "Alice",
                hand = mutableListOf(
                    Card(Suit.RED, 10),
                    Card(Suit.GREEN, 7)
                )
            )

            val currentTrick = listOf(
                Pair(PlayerState("player2", "Bob", mutableListOf()), Card(Suit.RED, 5))
            )

            assertFalse(
                TrickRules.isValidPlay(player, Card(Suit.GREEN, 7), currentTrick),
                "Player should not be allowed to play GREEN when holding RED"
            )

            assertTrue(
                TrickRules.isValidPlay(player, Card(Suit.RED, 10), currentTrick),
                "Player should be allowed to play RED (lead suit)"
            )
        }
    }

    @Nested
    inner class TrumpEdgeCases {
        @Test
        fun `higher trump beats lower trump`() {
            val playedCards = listOf(
                alice to Card(Suit.BLUE, 5),
                bob to Card(Suit.BLUE, 10),
                charlie to Card(Suit.RED, 13)
            )

            assertEquals(bob, TrickRules.determineTrickWinner(playedCards, trumpSuit = Suit.BLUE))
        }

        @Test
        fun `trump beats lead suit even when lower value`() {
            val playedCards = listOf(
                alice to Card(Suit.RED, 13),
                bob to Card(Suit.BLUE, 1),
                charlie to Card(Suit.RED, 10)
            )

            assertEquals(bob, TrickRules.determineTrickWinner(playedCards, trumpSuit = Suit.BLUE))
        }
    }
}